app/helpers/foreman_ansible/hosts_helper_extensions.rb

```diff
@@ -13,22 +13,34 @@ def ansible_roles_present?(host)
         host.inherited_ansible_roles.present?
     end
 
-    def ansible_roles_button(host)
-      link_to(
-        icon_text('play', ' ' + _('Ansible roles'), :kind => 'fa'),
-        play_roles_host_path(:id => host.id),
-        :id => :ansible_roles_button,
-        :class => 'btn btn-default',
-        :'data-no-turbolink' => true
-      )
+    def ad_hoc_role_button
+      [link_to_function(
+        icon_text('play', ' ' + _('Ad hoc Ansible role'), :kind => 'fa'),
+        'show_ad_hoc_role_modal()',
+        :id => :ansible_ad_hoc_role_button,
+        :class => 'btn btn-default'
+      )]
     end
 
-    def host_title_actions_with_run_ansible_roles(*args)
-      host = args.first
+    def ansible_roles_button(host)
       if ansible_roles_present?(host)
-        button = ansible_roles_button(host)
-        title_actions(button_group(button))
+        [link_to(
+          icon_text('play', ' ' + _('Ansible roles'), :kind => 'fa'),
+          play_roles_host_path(:id => host.id),
+          :id => :ansible_roles_button, :class => 'btn btn-default',
+          :'data-no-turbolink' => true
+        )]
+      else
+        []
       end
+    end
+
+    def host_title_actions_with_run_ansible_roles(*args)
+      buttons = ad_hoc_role_button
+      buttons += ansible_roles_button(args.first)
+
+      title_actions(button_group(buttons))
```

Comment:
Maybe you could put the buttons into dropdown?

![](http://i.imgur.com/QNAz0Ox.png)